src/content/blog/2022-08-03-new-config-system-part-1.md

```diff
@@ -0,0 +1,130 @@
+
+Over the years, however, the config system grew into an unwieldy mess. That's why in 2019 I proposed creating a new config system to make it easier to configure ESLint in a world where JavaScript projects are growing increasingly complex. A significant portion of the new config system has been merged into the main branch, and so it's time to start learning about how you will configure ESLint in the future. But in order to do that, it's helpful to take a look back and see how we got into the current state of things.
+
+## Incremental changes leading to maximum complexity
+
+Looking back at how the current config system (called the eslintrc system) evolved, every step made logical sense for where we were at the time. ESLint has always operated on an incremental approach to development where we look at ways to improve what we already have instead of throwing things away to start over. The eslintrc system was no different.
+
+### The `extends` key
+
+The first significant change to eslintrc was with the introduction of the `extends` key. The `extends` key, borrowed lovingly from JSHint, allowed users to import another configuration and then augment it, for example:
+
+```json
+{
+  "extends": ["./other-config.json"],
+  "rules": {
+    "semi": "warn"
+  }
+}
+```
+
+So assuming `./other-config.json` had some configuration data, you could import that and then add your own `rules` settings on top of it. This turned out to a great step forward for ESLint for a number of reasons.
+
+First, `extends` actually preceded the idea of shareable configs that could be distributed via npm. It was during the implementation of `extends` that we realized shareable configs were possible. The files specified in `extends` were loaded via the Node.js `require()` function, so anything Node.js could load through that function could also work as a config to extend from.
+
+Second, `extends` allowed us to implement `eslint:recommended`, the set of rules that we felt were important for everyone to enable. Originally ESLint has several rules enabled by default, but that became a burden for users. So we switched to having all rules off by default, which also was confusing for new users who didn't see any rules. Adding `eslint:recommended` allowed us to make it explicit that you were including a bunch of rules we recommended but you could remove them if you didn't want to.
+
+In hindsight, if we had thought things through a little more, we would have removed the configuration cascade at this point. Introducing `extends` enabled a lot of the same use cases as the cascade, and keeping both turned out to be a mess that we would spend years trying to fix.
+
+### Personal configs
+
+The next layer of complexity was added when people requested that we add the abilty to have a personal config file at `~/.eslintrc`. So we added an additional check: if we didn't find a config file in the ancestry of the file location, then we would automatically look for a personal config file.
```

Comment:
If I'm not mistaken, global configs were added way before `extends`, but it doesn't really matter.

Reply:
That is probably correct -- I was having a hard time tracking down when certain things were added, but agree, it doesn't really matter to the story.